### PR TITLE
Integration restart manager

### PR DESCRIPTION
**Pull request by trasc**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:
- Make the integration test framework able to stop the manager running in an envtest instance.
- Reuse the envtest instances in the test suites.
- Reorganize some suites for better performance while running in parallel.

#### Which issue(s) this PR fixes:

Relates to #2993 

#### Does this PR introduce a user-facing change?

```release-note
None
```

### CODE REVIEW

**Comment by mimowo** [comment]
> * Make the integration test framework able to stop the manager running in an envtest instance.
> * Reuse the envtest instances in the test suites.
> * Reorganize some suites for better performance while running in parallel.

Would you mind splitting this into dedicated PRs for the ease of reviewing?

**Comment by trasc** [comment]
> > * Make the integration test framework able to stop the manager running in an envtest instance.
> > * Reuse the envtest instances in the test suites.
> > * Reorganize some suites for better performance while running in parallel.
> 
> Would you mind splitting this into dedicated PRs for the ease of reviewing?

I'd prefer not to, the only file that contains "substantial" changes is [test/integration/framework/framework.go](https://github.com/kubernetes-sigs/kueue/pull/3054/files#diff-9be12d69f919437ea6abbe1fbcae80e1e97b5e28fdcbc515596187f25b3b6fe4), for the rest of the , it's just about changing the place where `fwk.Init()` is called (`BeforeSuite` instead of `BeforeAll` or similar) and for the webhook related tests moved in one package.

**Comment by mimowo** [comment]
> I'd prefer not to, the only file that contains "substantial" changes is [test/integration/framework/framework.go](https://github.com/kubernetes-sigs/kueue/pull/3054/files#diff-9be12d69f919437ea6abbe1fbcae80e1e97b5e28fdcbc515596187f25b3b6fe4), for the rest of them , it's just about changing the place where fwk.Init() is called (BeforeSuite instead of BeforeAll or similar) and for the webhook related tests moved in one package.

You can generally expect much faster reviews for PRs which are small and separate concerns, please read:
- https://www.kubernetes.dev/docs/guide/pull-requests/#smaller-is-better-small-commits-small-pull-requests
- https://google.github.io/eng-practices/review/reviewer/speed.html#large

In particular, moving the webhooks feels like separate PR, we would also maybe be able to see what is the performance benefit of various improvements separately. Otherwise it is hard to tell if one optimization helps and how much, if there is another optimization biasing the stats. 

Anyway, if you want I can try to review the PR in this form, but seeing that the process may not be as effective as it could.

**Comment by trasc** [comment]
> > I'd prefer not to, the only file that contains "substantial" changes is [test/integration/framework/framework.go](https://github.com/kubernetes-sigs/kueue/pull/3054/files#diff-9be12d69f919437ea6abbe1fbcae80e1e97b5e28fdcbc515596187f25b3b6fe4), for the rest of them , it's just about changing the place where fwk.Init() is called (BeforeSuite instead of BeforeAll or similar) and for the webhook related tests moved in one package.
> 
> You can generally expect much faster reviews for PRs which are small and separate concerns, please read:
> 
> * https://www.kubernetes.dev/docs/guide/pull-requests/#smaller-is-better-small-commits-small-pull-requests
> * https://google.github.io/eng-practices/review/reviewer/speed.html#large
> 
> In particular, moving the webhooks feels like separate PR, we would also maybe be able to see what is the performance benefit of various improvements separately. Otherwise it is hard to tell if one optimization helps and how much, if there is another optimization biasing the stats.
> 
> Anyway, if you want I can try to review the PR in this form, but seeing that the process may not be as effective as it could.

Most of the changes are trivial. Also the webhook moving needs to be in the same PR as the `Init` call changes, otherwise we'll have a broken intermediate state.

File: test/integration/framework/framework.go [hunk 1]
```diff
@@ -124,46 +127,71 @@ func (f *Framework) SetupClient(cfg *rest.Config) (context.Context, client.Clien
...
+				}
+				conn.Close()
+				return nil
+			}).Should(gomega.Succeed())
+		}
+	})
+}
+
+func (f *Framework) StopManager(ctx context.Context) {
+	ginkgo.By("stoping the manager", func() {
```
Comment (mimowo) [inline test/integration/framework/framework.go:182]:
```suggestion
	ginkgo.By("stopping the manager", func() {
```

File: test/integration/multikueue/suite_test.go [hunk 2]
```diff
@@ -96,7 +96,11 @@ func createCluster(setupFnc framework.ManagerSetup, apiFeatureGates ...string) c
 		APIServerFeatureGates: apiFeatureGates,
 	}
 	c.cfg = c.fwk.Init()
-	c.ctx, c.client = c.fwk.RunManager(c.cfg, setupFnc)
+	c.ctx, c.client = c.fwk.SetupClient(c.cfg)
+	// skip the manager setup if setup func is not provided
+	if setupFnc != nil {
```
Comment (mimowo) [inline test/integration/multikueue/suite_test.go:101]: Is this needed for the PR?

File: test/integration/scheduler/podsready/scheduler_test.go [hunk 3]
```diff
@@ -190,7 +185,7 @@ var _ = ginkgo.Describe("SchedulerWithWaitForPodsReady", func() {
 				Obj()
 			gomega.Expect(k8sClient.Create(ctx, testCQ)).Should(gomega.Succeed())
 			defer func() {
-				gomega.Expect(util.DeleteObject(ctx, k8sClient, testCQ)).Should(gomega.Succeed())
+				util.ExpectObjectToBeDeleted(ctx, k8sClient, testCQ, true)
```
Comment (mimowo) [inline test/integration/scheduler/podsready/scheduler_test.go:188]: I don't see how this change is related. Please move to a dedicated PR if not needed here.

File: test/integration/scheduler/podsready/scheduler_test.go [hunk 4]
```diff
@@ -161,9 +155,10 @@ var _ = ginkgo.Describe("SchedulerWithWaitForPodsReady", func() {
 			ginkgo.By("checking the first prod workload gets admitted while the second is waiting")
 			prodWl := testing.MakeWorkload("prod-wl", ns.Name).Queue(prodQueue.Name).Request(corev1.ResourceCPU, "2").Obj()
 			gomega.Expect(k8sClient.Create(ctx, prodWl)).Should(gomega.Succeed())
+			util.ExpectWorkloadsToHaveQuotaReservation(ctx, k8sClient, prodClusterQ.Name, prodWl)
```
Comment (mimowo) [inline test/integration/scheduler/podsready/scheduler_test.go:158]: I don't see how this change is related. Please move to a dedicated PR if not needed here.
.

File: test/integration/scheduler/podsready/scheduler_test.go [hunk 5]
```diff
@@ -650,7 +641,7 @@ var _ = ginkgo.Describe("SchedulerWithWaitForPodsReadyNonblockingMode", func() {
 				Obj()
 			gomega.Expect(k8sClient.Create(ctx, testCQ)).Should(gomega.Succeed())
 			defer func() {
-				gomega.Expect(util.DeleteObject(ctx, k8sClient, testCQ)).Should(gomega.Succeed())
+				util.ExpectObjectToBeDeleted(ctx, k8sClient, testCQ, true)
```
Comment (mimowo) [inline test/integration/scheduler/podsready/scheduler_test.go:644]: I don't see how this change is related. Please move to a dedicated PR if not needed here.

File: test/integration/multikueue/suite_test.go [hunk 6]
```diff
@@ -222,27 +226,26 @@ func managerAndMultiKueueSetup(ctx context.Context, mgr manager.Manager, gcInter
...
 	version, err := versionutil.ParseGeneric(os.Getenv("ENVTEST_K8S_VERSION"))
 	if err != nil || !version.LessThan(versionutil.MustParseSemantic("1.30.0")) {
 		managerFeatureGates = []string{"JobManagedBy=true"}
 	}
 
-	managerTestCluster = createCluster(func(ctx context.Context, mgr manager.Manager) {
-		managerAndMultiKueueSetup(ctx, mgr, gcInterval)
-	}, managerFeatureGates...)
+	// pass nil setup since the manager for the manage cluster is different in some specs.
+	managerTestCluster = createCluster(nil, managerFeatureGates...)
```
Comment (mimowo) [inline test/integration/multikueue/suite_test.go:237]: I don't think I follow, why the kueue manager setup is moved from here, but the kueue manager for workers stays here? I think it will be easier to understand if both types of manager where setup in the same place.

File: test/integration/webhook/jobs/mpijob_webhook_test.go [hunk 7]
```diff
@@ -23,24 +23,21 @@ import (
...
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 
+	"sigs.k8s.io/kueue/pkg/controller/jobs/mpijob"
 	testingjob "sigs.k8s.io/kueue/pkg/util/testingjobs/mpijob"
-	"sigs.k8s.io/kueue/test/integration/framework"
 	"sigs.k8s.io/kueue/test/util"
 )
 
-var _ = ginkgo.Describe("MPIJob Webhook", func() {
+var _ = ginkgo.Describe("MPIJob Webhook", ginkgo.Ordered, func() {
```
Comment (mimowo) [inline test/integration/webhook/jobs/mpijob_webhook_test.go:31]: Whys it the structure for the Describe and When changed (the Ordered label moved from When to Describe)? Also, I don't see ContinueOnFailure anymore. Is it needed in this PR?

**Review by mimowo — COMMENTED** [review]
Left some comments. TBH I'm still not sure why we cannot move the webhooks in a preparatory PR , commonizing the setup for webhooks first, but let it be.

File: test/integration/multikueue/suite_test.go (hunk 2 above)
Comment (trasc) [reply test/integration/multikueue/suite_test.go:101]: Yes , the worker clusters are fully initialized (including the managers) in before suite, the manager's manager is started/stopped per spec.

File: test/integration/multikueue/suite_test.go (hunk 6 above)
Comment (trasc) [reply test/integration/multikueue/suite_test.go:237]: There is no point in restarting the managers in the workers as long no configuration change is needed.

File: test/integration/scheduler/podsready/scheduler_test.go (hunk 4 above)
Comment (trasc) [reply test/integration/scheduler/podsready/scheduler_test.go:158]: This and a couple of other small fixes ware added to solve some flakes observed after switching to cluster reuse.  

For this case, historically the test was passing due to a higher load at the beginning due to envtest initialization.
Now any of the two can get admitted first and by moving this line we make prodWl is admitted first.

File: test/integration/scheduler/podsready/scheduler_test.go (hunk 3 above)
Comment (trasc) [reply test/integration/scheduler/podsready/scheduler_test.go:188]: This fixes an improper cleanup which used to be unnoticed because the envtest cluster used to be replaced.

File: test/integration/scheduler/podsready/scheduler_test.go (hunk 5 above)
Comment (trasc) [reply test/integration/scheduler/podsready/scheduler_test.go:644]: same

File: test/integration/webhook/jobs/mpijob_webhook_test.go (hunk 7 above)
Comment (trasc) [reply test/integration/webhook/jobs/mpijob_webhook_test.go:31]: In our test `ginkgo.Ordered` is label to decorate a container in which we need to be able to call `ginkgo.BeforeAll` (especially in this Spec which has a single `It`, it has no other value),  The `BeforeAll` was moved just for to me this look similar to the other webhook specs.

File: test/integration/framework/framework.go (hunk 1 above)
Comment (trasc) [reply test/integration/framework/framework.go]: done

File: test/integration/scheduler/podsready/scheduler_test.go (hunk 4 above)
Comment (mimowo) [reply test/integration/scheduler/podsready/scheduler_test.go:158]: if these are fixes to flakes, then it makes sense to merge them separately so that we can have the release branch more stable? I see that it didn't fail before (almost never), but the point holds, IIUC, that the code was buggy.

File: test/integration/multikueue/suite_test.go (hunk 6 above)
Comment (mimowo) [reply test/integration/multikueue/suite_test.go:237]: I see, ok

File: test/integration/webhook/jobs/mpijob_webhook_test.go (hunk 7 above)
Comment (mimowo) [reply test/integration/webhook/jobs/mpijob_webhook_test.go:31]: sgtm, though I prefer to see drive-by cleanups in a separate PR to reduce distractions. Anyway, I can accept it after I now understand.

File: test/integration/multikueue/suite_test.go (hunk 2 above)
Comment (mimowo) [reply test/integration/multikueue/suite_test.go:101]: ok

**Comment by mimowo** [comment]
I would prefer some of the drive-by fixes/cleanups to be in dedicated PRs so that they could cherry-pick if needed. Also, to reduce distractions and minimize the cognitive load. As mentioned in https://github.com/kubernetes-sigs/kueue/pull/3054#issuecomment-2360285542, I believe it makes the review process easier and faster overall.

In any case, I think I understand the changes now, and are ok to go with them.

/lgtm
/approve

**Comment by trasc** [comment]
1/5
/test pull-kueue-test-integration-main

**Comment by trasc** [comment]
2/5
/test pull-kueue-test-integration-main

**Comment by trasc** [comment]
3/5
/test pull-kueue-test-integration-main

**Comment by trasc** [comment]
4/5
/test pull-kueue-test-integration-main

**Comment by trasc** [comment]
5/5
/test pull-kueue-test-integration-main

**Comment by trasc** [comment]
5 out of 5 ok, @mimowo please have another look
